File: Task-002/wrdcount.scala
//Frederick Herzog
// Scala Word Count 
// 11/3/2020

import scala.io.Source
import java.io.File
import java.io.PrintWriter
import scala.util.matching.Regex
import scala.collection.immutable.ListMap
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.{FileSystem, Path}

object wrdcount {
  libraryDependencies ++= Seq(
  "org.apache.hadoop" % "hadoop-client" % "2.7.3"
)
	var file_path = "Shakespeare.txt"
  var out_path = "count_scala.txt"
  var dfs_path = "hdfs://localhost:9000/SPtext/"+out_path

   def main(args: Array[String]) {
       var dat = readData(file_path)
       var my_count = countWords(dat)
       printf("%s", my_count)
       val sorted_map = sort_Map(my_count)
       printMap(sorted_map)
       write_to_File(sorted_map, out_path)
       write_to_dfs(out_path, dfs_path)
   }
	
	def readData(path: String): List[String] = {
		// Using Regex to filter out unwanted characters.
       val words = """([A-Za-z0-9])+""".r
        // Reading each word in the file into a List using flatMap.
       val wordList = io.Source.fromFile(path).getLines.flatMap(words.findAllIn).toList
       val lc_wordList = convert_to_lowerCase(wordList)
       return lc_wordList
        }
    
    def convert_to_lowerCase(dat: List[String]): List[String] = {
    	// Convert each word in List to lowercase.
    	val lowerCase = for (w <- dat) yield w.toLowerCase
    	return lowerCase
    }
    
    def countWords(dat: List[String]): Map[String, Int] = {
    	// Using a map to count the number of occurences for each word.
    	dat.toSet.map((word: String) => (word, dat.count(_ == word))).toMap
    	}

    def sort_Map(dat: Map[String, Int]): Map[String, Int] = {
    	// Sorting the map (descending order) using ListMap.
    	return ListMap(dat.toSeq.sortWith(_._2 < _._2):_*)
    }

    def printMap(dat: Map[String, Int]) = {
    	// Printing out the data from the map.
    	for ((k,v) <- dat) printf("%s : %s\n", k, v)
    }

    def write_to_File(dat: Map[String, Int], path: String) = {
      val writer = new PrintWriter(new File(path))
      dat.foreach {
        case (k, v) => writer.write(k + "  " + v + "\n")

    }
      writer.close()

  }

    def write_to_dfs(path: String, dfs_path: String, dat: Map[String, Int]) = {
      val path = new Path(path)
      val conf = new Configuration
      conf.set("fs.defaultFS", dfs_path)
      val fs = FileSystem.get(conf)
      val out = fs.create(path)
      out.write(dat)
      fs.close()
  }

}

File: Task-002/wrdcount.py
#Frederick Herzog
# Python Word Count
# 11/3/2020

from re import sub
from collections import Counter
from pydoop import hdfs

def txt_WordsToList(file):
	with open(file, 'r') as f:
		data = f.read().split()
		return data
	
def removePunct(dat):
	cleandat = []
	for word in dat:
		cleaned_word = sub('[^A-Za-z0-9]+', '', word)
		cleandat.append(cleaned_word)
	return cleandat

def makeLowerCase(dat):
	lower_case = [w.lower() for w in dat]
	return lower_case

def countWords(dat):
	count = Counter
	c = count(dat)
	return dict_to_tuples(c)

def dict_to_tuples(dat):
	freq = []
	for k, v in dat.items():
		freq.append((v,k))
	freq.sort(reverse = True)
	return freq

def out_to_txt(dat, file):
	with open(file, 'w') as f:
		for i in dat:
			f.write('    '.join (str(s) for s in i) + '\n')

def out_to_dfs(file, dfs_path):
	print("Writing file to HDFS...")
	hdfs.put(file, dfs_path)

if __name__ == '__main__':
	file_path = "Shakespeare.txt"
	out_path = "count_python.txt"
	dfs_path = "hdfs://localhost:9000/SPtext/"+out_path
	words = txt_WordsToList(file_path)
	words_no_punct = removePunct(words)
	lower_c_words = makeLowerCase(words_no_punct)
	my_count = countWords(lower_c_words)
	print(my_count)
	out_to_txt(my_count, out_path)
	out_to_dfs(out_path, dfs_path)



